examples/weather_station/src/app_task.cpp

```diff
@@ -339,6 +363,121 @@ void AppTask::UpdateClusterState()
 	}
 }
 
+void AppTask::UpdatePowerSourceClusterState() {
+	EmberAfStatus status;
+	uint32_t voltage = BatteryMeasurementRead();
+	/* Value is expressed in half percent units ranging from 0 to 200. */
+	uint8_t batteryPercentage;
+	EmberAfPowerSourceStatus batteryStatus;
+	EmberAfBatChargeLevel batteryChargeLevel;
+	bool batteryPresent;
+	EmberAfBatChargeState batteryCharged;
+
+	if (voltage < 0) {
+		voltage = 0;
+		batteryPercentage = 0;
+		batteryStatus = EMBER_ZCL_POWER_SOURCE_STATUS_UNAVAILABLE;
+		batteryPresent = false;
+
+		LOG_ERR("Battery level measurement failed %d", voltage);
+
+	} else {
+		batteryStatus = EMBER_ZCL_POWER_SOURCE_STATUS_ACTIVE;
+		batteryPresent = true;
+	}
+
+	if (voltage <= kMinimalOperatingVoltageMv) {
+		batteryPercentage = kMinBatteryPercentage;
+	} else if (voltage >= kMaximalOperatingVoltageMv) {
+		batteryPercentage = kMaxBatteryPercentage;
+	} else {
+		batteryPercentage = kMaxBatteryPercentage * (voltage - kMinimalOperatingVoltageMv) / (kMaximalOperatingVoltageMv - kMinimalOperatingVoltageMv);
+	}
+
+	if ((voltage < kWarningThresholdVoltageMv) && (voltage >= kCriticalThresholdVoltageMv)) {
```

Comment:
If you swapped the first two `if`s you would need to check `kCriticalThresholdVoltageMv` twice.

Reply:
Very nice, thanks!